### PR TITLE
*: use gRPC server GracefulStop

### PR DESCRIPTION
**Pull request by gyuho**
Example output

```
2017-04-14 09:20:53.599484 I | etcdserver/api: enabled capabilities for version 3.2
^C2017-04-14 09:20:55.974829 N | pkg/osutil: received interrupt signal, shutting down...
2017-04-14 09:20:55.974857 I | etcdserver: skipped leadership transfer for single member cluster
2017-04-14 09:20:55.974875 W | embed: gracefully stopping gRPC server
2017-04-14 09:20:55.974883 W | embed: gracefully stopped gRPC server
2017-04-14 09:20:55.975058 W | embed: server stopped with "grpc: the server has been stopped"
2017-04-14 09:20:55.975482 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975503 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975523 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975540 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975559 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975598 N | embed: serving insecure client requests on 127.0.0.1:2379, this is strongly discouraged!
2017-04-14 09:20:55.975628 W | embed: server stopped with "accept tcp 127.0.0.1:2379: use of closed network connection"
2017-04-14 09:20:55.975667 W | embed: server stopped with "mux: listener closed"
2017-04-14 09:20:55.975712 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975729 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
2017-04-14 09:20:55.975743 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2379: getsockopt: connection refused"; Reconnecting to {127.0.0.1:2379 <nil>}
```

Fix https://github.com/coreos/etcd/issues/7322.

### CODE REVIEW

**Comment by heyitsanthony** [comment]
This should fix the inflight op crashes, right? Can there be a test?

**Comment by gyuho** [comment]
@heyitsanthony Yes, I will try to verify this fixes the problem by adding tests or reproduce.

**Comment by gyuho** [comment]
@heyitsanthony Test added. Confirmed that it fixes the issue (use `reqN := 500` and comment out graceful stop part, then it will panic in boltdb). PTAL.

File: embed/serve.go [hunk 1]
```diff
@@ -52,11 +52,12 @@ type serveCtx struct {
 
 	userHandlers    map[string]http.Handler
 	serviceRegister func(*grpc.Server)
+	stopGRPCc       chan func()
```
Comment (heyitsanthony) [inline embed/serve.go]: `grpcServers []*gprc.Server`

File: embed/serve.go [hunk 2]
```diff
@@ -74,6 +75,12 @@ func (sctx *serveCtx) serve(s *etcdserver.EtcdServer, tlscfg *tls.Config, handle
 
 	if sctx.insecure {
 		gs := v3rpc.Server(s, nil)
+		sctx.stopGRPCc <- func() {
```
Comment (heyitsanthony) [inline embed/serve.go]: sctx.grpcServers = sctx.append(sctx.grpcServers, gs)

File: embed/serve.go [hunk 3]
```diff
@@ -103,6 +110,12 @@ func (sctx *serveCtx) serve(s *etcdserver.EtcdServer, tlscfg *tls.Config, handle
 
 	if sctx.secure {
 		gs := v3rpc.Server(s, tlscfg)
+		sctx.stopGRPCc <- func() {
```
Comment (heyitsanthony) [inline embed/serve.go]: sctx.grpcServers = sctx.append(sctx.grpcServers, gs)

File: etcdserver/config.go [hunk 4]
```diff
@@ -61,6 +61,12 @@ type ServerConfig struct {
 	ClientCertAuthEnabled bool
 
 	AuthToken string
+
+	// OnShutdown gracefully stops gRPC server on shutdown.
```
Comment (heyitsanthony) [inline etcdserver/config.go]: this should be a generic thing instead of talking about grpc, etc

`// OnShutdown is called immediately before releasing etcd server resources.`

File: integration/v3_grpc_inflight_test.go [hunk 5]
```diff
@@ -75,3 +76,38 @@ func TestV3MaintenanceDefragmentInflightRange(t *testing.T) {
 
 	<-donec
 }
+
```
Comment (heyitsanthony) [inline integration/v3_grpc_inflight_test.go]: there were some changes to the mvcc code so `TestV3MaintenanceHashInflight` would appear to work. Namely,  `TestStoreHashAfterForceCommit` and the `stopc` logic in `Hash` should probably be removed.

File: integration/v3_grpc_inflight_test.go [hunk 6]
```diff
@@ -75,3 +76,38 @@ func TestV3MaintenanceDefragmentInflightRange(t *testing.T) {
...
+func TestV3KVInflightRangeRequests(t *testing.T) {
+	defer testutil.AfterTest(t)
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	cli := clus.RandClient()
+	kvc := toGRPC(cli).KV
+
+	if _, err := kvc.Put(context.Background(), &pb.PutRequest{Key: []byte("foo"), Value: []byte("bar")}); err != nil {
+		panic(err)
```
Comment (heyitsanthony) [inline integration/v3_grpc_inflight_test.go]: t.Fatal(err)

File: embed/etcd.go [hunk 7]
```diff
@@ -137,6 +139,11 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 	if err = e.serve(); err != nil {
 		return
 	}
+	e.Server.Cfg.OnShutdown = func() {
```
Comment (heyitsanthony) [inline embed/etcd.go]:
```go
OnShutdown = func() {
    for _, sctx := range e.sctxs {
        for _, gs := range sctx.grpcServers {
            gs.GracefulStop()
        }
    }
}
```

File: embed/etcd.go [hunk 8]
```diff
@@ -343,6 +350,10 @@ func (e *Etcd) serve() (err error) {
 }
 
 func (e *Etcd) errHandler(err error) {
+	if transport.IsClosedConnError(err) || err == grpc.ErrServerStopped {
```
Comment (heyitsanthony) [inline embed/etcd.go]: why is this necessary? shouldn't `stopc` be closed before calling etcdserver.Stop?

File: etcdserver/server.go [hunk 9]
```diff
@@ -689,6 +689,12 @@ func (s *EtcdServer) run() {
 	}
 
 	defer func() {
+		// stop accepting new connections, RPCs,
+		// and blocks until all pending RPCs are finished
+		if s.Cfg != nil && s.Cfg.OnShutdown != nil {
+			s.Cfg.OnShutdown()
```
Comment (heyitsanthony) [inline etcdserver/server.go]: I think it's possible to avoid having this in Cfg entirely-- this function could be called prior to calling HardStop/Stop; similar to how the listeners are closed in embed.Etcd.Close() before calling `Server.Stop()`

Comment (gyuho) [inline etcdserver/server.go]: Ok let me re-organize the code.

File: embed/etcd.go (hunk 8 above)
Comment (gyuho) [inline embed/etcd.go]:
```go
// in embed/etcd.go
func StartEtcd(inCfg *Config) (e *Etcd, err error) {
	if err = inCfg.Validate(); err != nil {
		return nil, err
	}
	e = &Etcd{cfg: *inCfg, stopc: make(chan struct{})}
	cfg := &e.cfg
	defer func() {
		if e != nil && err != nil {
			e.Close()
			e = nil
		}
	}()
```

We close `stopc` by calling `e.Close()` here, but `StartEtcd` returns with `nil` error, so it is not called in our use case?

File: embed/etcd.go (hunk 7 above)
Comment (gyuho) [inline embed/etcd.go]: I think this does not sync with `serve` routine? We populate `sctx.grpcServers`
in `(e *Etcd) serve()` (which calls `(sctx *serveCtx) serve` that creates `*grpc.Server`.
But `(e *Etcd) serve()` returns calling `(sctx *serveCtx) serve` in goroutines.

**Comment by gyuho** [comment]
I think we still need `GracefulStop` in `etcdserver` because `embed.Etcd.Close` won't be triggered unless there's an error at the beginning of serving. And `etcdserver.EtcdServer.Stop` is the handler registered for OS interrupt signals.

```go
// etcdmain/etcd.go

// startEtcd runs StartEtcd in addition to hooks needed for standalone etcd.
func startEtcd(cfg *embed.Config) (<-chan struct{}, <-chan error, error) {
	if cfg.Metrics == "extensive" {
		grpc_prometheus.EnableHandlingTimeHistogram()
	}

	e, err := embed.StartEtcd(cfg)
	if err != nil {
		return nil, nil, err
	}
	osutil.RegisterInterruptHandler(e.Server.Stop)
```

File: embed/etcd.go (hunk 7 above)
Comment (gyuho) [inline embed/etcd.go]: nvm, we somehow need a way to sync to slice anyway.

**Comment by heyitsanthony** [comment]
why not have `osutil.RegisterInterruptHandler(e.Close)`?

File: embed/etcd.go [hunk 10]
```diff
@@ -147,6 +161,7 @@ func (e *Etcd) Config() Config {
 
 func (e *Etcd) Close() {
 	e.closeOnce.Do(func() { close(e.stopc) })
+	e.OnShutdown()
```
Comment (heyitsanthony) [inline embed/etcd.go]: can the function be inlined here instead of needing a separate `OnShutdown` field?

File: embed/etcd.go [hunk 11]
```diff
@@ -137,6 +140,17 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 	if err = e.serve(); err != nil {
 		return
 	}
+	e.OnShutdown = func() {
+		// (gRPC server) stops accepting new connections,
+		// RPCs, and blocks until all pending RPCs are finished
+		for _, sctx := range e.sctxs {
+			for gs := range sctx.grpcServerC {
+				plog.Warning("gracefully stopping gRPC server")
```
Comment (heyitsanthony) [inline embed/etcd.go]: don't warn / print anything? this should be part of the normal shutdown process

File: mvcc/backend/batch_tx.go [hunk 12]
```diff
@@ -166,19 +166,12 @@ func (t *batchTx) commit(stop bool) {
...
-			//   2. batchTx.commit(true) from stopping backend
-			//   3. batchTx.commit(false) from inflight mvcc Hash call
-			//
-			// Check if db is nil to prevent this panic
-			if t.tx.DB() != nil {
-				atomic.StoreInt64(&t.backend.size, t.tx.Size())
-			}
+			// t.tx.DB()==nil if 'CommitAndStop' calls 'batchTx.commit(true)',
+			// which initializes *bolt.Tx.db and *bolt.Tx.meta as nil; panics t.tx.Size().
+			// Server must make sure 'batchTx.commit(false)' does not follow
```
Comment (heyitsanthony) [inline mvcc/backend/batch_tx.go]: This probably shouldn't mention the etcd server or gRPC. The contract is independent of all that-- don't have any operations inflight when closing the backend.

File: integration/v3_grpc_inflight_test.go [hunk 13]
```diff
@@ -15,63 +15,79 @@
...
+			if err != nil && grpc.ErrorDesc(err) != context.Canceled.Error() {
+				t.Fatalf("inflight request should be canceld with %v, got %v", context.Canceled, err)
+			}
+		}()
+	}
 
-	mvc := toGRPC(cli).Maintenance
-	mvc.Defragment(context.Background(), &pb.DefragmentRequest{})
+	// simulate 'embed.Etcd.Close()' with '*grpc.Server.GracefulStop'
+	clus.Members[0].grpcServer.GracefulStop()
```
Comment (heyitsanthony) [inline integration/v3_grpc_inflight_test.go]: `clus.Members[0].Stop()`

File: mvcc/kvstore_test.go [hunk 14]
```diff
@@ -518,20 +518,6 @@ func newTestKeyBytes(rev revision, tombstone bool) []byte {
 	return bytes
 }
 
-// TestStoreHashAfterForceCommit ensures that later Hash call to
-// closed backend with ForceCommit does not panic.
-func TestStoreHashAfterForceCommit(t *testing.T) {
```
Comment (heyitsanthony) [inline mvcc/kvstore_test.go]: also remove the `select` in `mvcc.store.Hash`, which was faking this

**Comment by FingerLiu** [comment]
@gyuho is there a workaround in early versions?
